### PR TITLE
Update to graphback 0.16.0 and refactor resolver definitions

### PR DESCRIPTION
**Pull request by wtrocki**
This is designed as long running branch to fix all remaining problems.

Targets:
- update graphback
- update apollo libraries and prepare client to use offix datastore

### CODE REVIEW

**Comment by wtrocki** [comment]
To replicate:
```
yarn
yarn generate:all  
```

**Comment by machi1990** [comment]
I managed to reproduce the issue https://github.com/aerogear/graphback/issues/2019 reported here and it cause of relationship `key` mismatch possibly introduced by https://github.com/aerogear/graphback/pull/1818 (I'll look more into that as part of https://github.com/aerogear/graphback/pull/2020). 

Can you try this schema 
```graphql
"""
@model(delete: false)
"""
type DistributionCentre implements Address {
  _id: GraphbackObjectID!
  name: String
  address1: String
  address2: String
  city: String
  postcode: Int
  lat: Float
  long: Float
  """
  @oneToMany(field: "distributionCentre")
  """
  products: [Product]

  """
  @oneToMany(field: "distributionCentre")
  """
  actions: [VolunteerAction]
}

"""
@model(delete: false)
"""
type Volunteer {
  _id: GraphbackObjectID!
  firstName: String
  lastName: String
  email: String
  username: String!
  address1: String
  address2: String
  city: String
  postcode: Int
  dateOfBirth: GraphbackDateTime
  canDeliver: Boolean
  actionsCompleted: Int
  actionsActive: Int
  active: Boolean

  """
  @oneToMany(field: 'volunteer')
  """
  actions: [VolunteerAction]
}

"""
@model(delete: false)
"""
type Recipient implements Address {
  _id: GraphbackObjectID!
  firstName: String
  lastName: String
  phone: String
  address1: String
  address2: String
  postcode: Int
  city: String
  lat: Float
  long: Float
  actionsCompleted: Int
  deliveryDays: String
  """
  Comma separated list of preferrend products labels.
  Easier to create volunteer actions when knowing what one really needs.
  """
  prefferedProducts: String
  """
  @oneToMany(field: 'recipient')
  """
  actions: [VolunteerAction]
}

"""
Represents action that is assigned to volunteer

@model
"""
type VolunteerAction {
  _id: GraphbackObjectID!
  title: String!
  description: String
  status: ActionStatus
  assignedAt: GraphbackDateTime
  completedAt: GraphbackDateTime
  _createdAt: GraphbackDateTime

  """
  @manyToOne(field: 'actions')
  """
  volunteer: Volunteer

  """
  @manyToOne(field: 'actions')
  """
  distributionCentre: DistributionCentre
  
  """
  @oneToMany(field: 'volunteerAction')
  """
  products: [VolunteerActionProduct]
}

"""
Represents a join model between a volunteer action and product

@model(create: true, update: false, delete: false, subCreate: false, subUpdate: false, subDelete: false)
"""
type VolunteerActionProduct {
  _id: GraphbackObjectID!
}

interface Address {
  address1: String
  address2: String
  city: String
  postcode: Int
  lat: Float
  long: Float
}

"""
@model(delete: false)
"""
type Product {
  _id: GraphbackObjectID!
  label: String!
  description: String
  """
  @oneToMany(field: 'product')
  """
  volunteerActionProducts: [VolunteerActionProduct]
}

"""
@model
"""
type DailyActionPlan {
  _id: GraphbackObjectID!
  owner: String
  date: GraphbackDateTime
  numberOfCasesCreated: Int
  numberOfVolunteersAssigned: Int
  numberOfRecipients: Int
}

enum ActionStatus {
  CREATED
  ASSIGNED
  COMPLETED
}

scalar GraphbackDateTime
scalar GraphbackObjectID

## Custom logic
type Mutation {
  """
  Automatically create actions for the day of the month specified
  """
  assignVolunteers: DailyActionPlan
}
``` 

To see if it improves things? When I tested with that schema I was able to perform the generation and start the server.

/cc @wtrocki @craicoverflow

**Comment by wtrocki** [comment]
Strange. I have the same and generates something wrong:

```
  """
  @manyToOne(field: 'actions', key: 'volunteerId')
  @manyToOne(field: 'actions')
  """
```
Then it fails to start.

I have changed mappyingHelpers;

```
        if (!relationshipAnnotation && relationshipAnnotation.kind !== 'manyToOne') {
```

This allowed me to remove manyToOne and get things running

**Comment by machi1990** [comment]
> Strange. I have the same and generates something wrong:
> 
> ```
>   """
>   @manyToOne(field: 'actions', key: 'volunteerId')
>   @manyToOne(field: 'actions')
>   """
> ```

This double generation is normal, and graphql-metadata would ignore the second annotation once it encounters the first one. 
 
> Then it fails to start.

can you double check? It should start just fine.

**Comment by wtrocki** [comment]
> This double generation is normal, and graphql-metadata would ignore the second annotation once it encounters the first one. 

From the point of the user viewing the schema it looks strange.

**Comment by machi1990** [comment]
> > This double generation is normal, and graphql-metadata would ignore the second annotation once it encounters the first one.
> 
> From the point of the user viewing the schema, it looks strange, but I get this is hard to do from the cli. I see that as regression as my schema did not had that before.
> 
> Screenshot 2020-09-10 at 15 18 52

Yes, when inspecting the schema this brings a bad and a regression in UX. The only explanation I could think of, as far as running the app is concerning it does not impact on running the app since the second annotation is completely ignore by Graphback and the first one is needed and generated by Graphback for a consistent relationship parsing. It's something we can look to improve without impacting performance much. 

A workaround would be to make sure that the relationship contains the `key` field and it has to match the corresponding inverse relationship `key` to avoid the validation error.

Happy to discuss this further here https://github.com/aerogear/graphback/issues/2023

**Comment by wtrocki** [comment]
@machi1990 I have small problem with the custom resolver. It seems that assignment is not possible.
Apart from that all works

**Comment by machi1990** [comment]
> @machi1990 I have small problem with the custom resolver. It seems that assignment is not possible.
> Apart from that all works. Would you have idea why this is failing (I haven't changed anything in terms of database etc.)

@wtrocki thank you for the great job. I'll have a look to see what's wrong with the custom resolver.

**Comment by machi1990** [comment]
@wtrocki I pushed https://github.com/aerogear/OpenVolunteerPlatform/pull/153/commits/ee9079689686735c9c5703c2f7f087a8ca46a6f6 that should fix the issue plus adds the ability to retrieve "DailyActionPlan" owner from the current authz context. Previously every automatic action was owned by "ovp-admin", now the owner will be the one who initiated the the scheduling i.e current connected user.

**Comment by wtrocki** [comment]
@machi1990 Thank you so much for taking over that part. I was little bit confused. Going to review it now.

**Comment by wtrocki** [comment]
> Previously every automatic action was owned by "ovp-admin", now the owner will be the one who initiated the the scheduling i.e current connected user.

Awesome!

File: platform/client-admin/.graphqlrc.yml [hunk 1]
```diff
@@ -31,10 +31,16 @@ extensions:
...
+            GraphbackDate: Date
             GraphbackDateTime: Date
+            GraphbackTimestamp: number
             GraphbackObjectID: string
+            GraphbackJSON: any
+            GraphbackJSONObject: "{ [key: string]: any }"
         plugins:
-          - add: '/* tslint:disable */'
+          - add: 
+              content: '/* tslint:disable */'
```
Comment (craicoverflow) [inline platform/client-admin/.graphqlrc.yml:43]:
```suggestion
              content: '/* eslint-disable */'
```

Still using tslint?

Comment (wtrocki) [reply platform/client-admin/.graphqlrc.yml:43]: Yep.. Same as starter. Starter needs investment. As for OVP it will not matter much

**Comment by machi1990** [comment]
> @machi1990 I have small problem with the custom resolver. It seems that assignment is not possible.
> Apart from that all works. Would you have idea why this is failing (I haven't changed anything in terms of database etc.)

The main issue with that I saw was that the `info: GraphQLResolveInfo` was passed to every service call which resulted in a "wrong" projection being sent to the MongoDB. When the `assignVolunteers` mutation is called, the `info`  parameter will contain resolvable fields for `DailyActionPlan` model and not any other business model meaning that once the query is sent to `db` it won't err (as a relational database would) but simply tries its best to project the db results to the requested fields. If this query is sent to retrieve a list of `Products`, it will simply return `_id`, as the other fields do not belong to the `product` collection but `dailyactionplan`.

File: platform/server/src/resolvers/custom-resolvers.ts [hunk 2]
```diff
@@ -1,161 +1,246 @@
...
+      }
+
+      const now = new Date();
+
+      // find volunteers and recipients with ongoing actions
+      // and store their ids somewhere so that they will not be picked as candidates
+      // for automatic scheduling
+      const { items } = await graphback.VolunteerAction.findBy(
+        { filter: { status: { in: ["CREATED", "ASSIGNED"] } } },
+        context
```
Comment (craicoverflow) [inline platform/server/src/resolvers/custom-resolvers.ts:31]: Pass info also to allow CRUD service to extract selected fields

Comment (machi1990) [reply platform/server/src/resolvers/custom-resolvers.ts:31]: See https://github.com/aerogear/OpenVolunteerPlatform/pull/153#issuecomment-690965297 for a longer explanation

File: platform/server/src/resolvers/custom-resolvers.ts [hunk 3]
```diff
@@ -1,161 +1,246 @@
...
-                distributionCentreId: {
-                    ne: undefined
+      // let's retrieve recipients who do not have action
+      const findRecipients = graphback.Recipient.findBy(
+        {
+          filter: buildQueryToFindAllItemsExcludingThoseWithGivenIds(
+            recipientIds
+          )
+        },
+        context
```
Comment (craicoverflow) [inline platform/server/src/resolvers/custom-resolvers.ts:59]: Pass info also to allow CRUD service to extract selected fields

File: platform/server/src/resolvers/custom-resolvers.ts [hunk 4]
```diff
@@ -1,161 +1,246 @@
...
+      }
 
+      // let's retrieve volunteers who are deemed free
+      const findVolunteers = graphback.Volunteer.findBy(
+        {
+          filter: buildQueryToFindAllItemsExcludingThoseWithGivenIds(
+            volunteerIds
+          )
+        },
+        context
```
[the remark this answers is not in the document]
Comment (machi1990) [reply platform/server/src/resolvers/custom-resolvers.ts:49]: See https://github.com/aerogear/OpenVolunteerPlatform/pull/153#issuecomment-690965297

File: platform/server/src/resolvers/custom-resolvers.ts (hunk 3 above)
Comment (machi1990) [reply platform/server/src/resolvers/custom-resolvers.ts:59]: See https://github.com/aerogear/OpenVolunteerPlatform/pull/153#issuecomment-690965297

**Comment by machi1990** [comment]
> > Previously every automatic action was owned by "ovp-admin", now the owner will be the one who initiated the the scheduling i.e current connected user.
> 
> Awesome! I thought only ovp-admin can login to admin side. I will play with this.

Well, `ovp-admin` is the only admin account we initialize the application with but a user can create their own admins. So let's make this as generic as possible.

**Comment by wtrocki** [comment]
> The main issue with that I saw was that the info: GraphQLResolveInfo was passed to every service call which resulted in a "wrong" projection being sent to the MongoDB. When the assignVolunteers mutation is called, the info parameter will contain resolvable fields for DailyActionPlan model and not any other business model meaning that once the query is sent to db it won't err (as a relational database would) but simply tries its best to project the db results to the requested fields. If this query is sent to retrieve a list of Products, it will simply return _id, as the other fields do not belong to the product collection but dailyactionplan.

This got me confused. I would blindlessly pass info to any resolver, not knowing the purpose of this. 
In my mind every call of the service needs to pass context, info (which was clearly wrong). 

I did read docs for the resolver but did not noticed this info on bottom.
Just used snippet from the top:

https://graphback.dev/docs/resolvers/custom-resolvers#info

**Comment by machi1990** [comment]
> > The main issue with that I saw was that the info: GraphQLResolveInfo was passed to every service call which resulted in a "wrong" projection being sent to the MongoDB. When the assignVolunteers mutation is called, the info parameter will contain resolvable fields for DailyActionPlan model and not any other business model meaning that once the query is sent to db it won't err (as a relational database would) but simply tries its best to project the db results to the requested fields. If this query is sent to retrieve a list of Products, it will simply return _id, as the other fields do not belong to the product collection but dailyactionplan.
> 
> This got me confused. I would blindlessly pass info to any resolver, not knowing the purpose of this.
> In my mind every call of the service needs to pass context, info (which was clearly wrong).
>

One of the reason I wanted to keep info out of the crud services to be honest. 

> I think it will be cool if we have example resolve that does one with and one without info - putting little more context when and where not pass it. No strong opinion, but got confused little bit.

Yeap. And probably feed in much in the api docs of the crud service (there is probably this info).

**Comment by wtrocki** [comment]
Works for me

**Review by machi1990 — APPROVED** [review]
LGTM.

Thanks @wtrocki for much of the work.

**Pull request merged**